extensions/amp-smartlinks/0.1/amp-smartlinks.js

```diff
@@ -0,0 +1,195 @@
+import {Linkmate} from './linkmate';
+import {getConfigOptions} from './linkmate-options';
+
+
+export class AmpSmartlinks extends AMP.BaseElement {
+  /**
+   * @param {!AmpElement} element
+   */
+  constructor(element) {
+    super(element);
+
+    /** @private {?../../../src/service/xhr-impl.Xhr} */
+    this.xhr_ = null;
+
+    /** @private {?../../../src/service/ampdoc-impl.AmpDoc} */
+    this.ampDoc_ = null;
+
+    /** @private {?./link-rewriter/link-rewriter-manager.LinkRewriterManager} */
+    this.linkRewriterService_ = null;
+
+    /** @private {?./link-rewriter/link-rewriter.LinkRewriter} */
+    this.smartLinkRewriter_ = null;
+
+    /** @private {?Object} */
+    this.linkmateOptions_ = null;
+
+    /** @private {?function({JsonObject})} */
+    this.fetchLinkmateConfig_ = null;
+
+    /** @private {?../../../src/service/viewer-impl.Viewer} */
+    this.viewer_ = null;
+
+    /** @private {?./linkmate.Linkmate} */
+    this.linkmate_ = null;
+
+    /** @private {?string} */
+    this.referrer_ = null;
+  }
+
+  /** @override */
+  buildCallback() {
+    this.xhr_ = Services.xhrFor(this.win);
+    this.ampDoc_ = this.getAmpDoc();
+
+    this.viewer_ = Services.viewerForDoc(this.ampDoc_);
+
+    this.linkmateOptions_ = getConfigOptions(this.element);
+    this.fetchLinkmateConfig_ = this.getLinkmateOptions_;
+
+    this.linkRewriterService_ = new LinkRewriterManager(this.ampDoc_);
+
+    return this.ampDoc_.whenBodyAvailable()
+        .then(() => this.viewer_.getReferrerUrl())
+        .then(referrer => {
+          this.referrer_ = referrer;
+          this.runSmartlinks_();
+        });
+  }
+
+  /**
+   * Wait for the config promise to resolve and then proceed to functionality
+   * @private
+   */
+  runSmartlinks_() {
+    this.fetchLinkmateConfig_().then(config => {
+      this.linkmateOptions_.linkmateExpected = config['linkmate_enabled'];
+      this.linkmateOptions_.publisherID = config['publisher_id'];
+
+      this.postPageImpression_();
+      this.linkmate_ = new Linkmate(
+          this.ampDoc_,
+          this.xhr_,
+          this.linkmateOptions_
+      );
+    });
+
+    this.smartLinkRewriter_ = this.initLinkRewriter_();
+
+    // We have a valid linkmate class, the config specified linkmate to run
+    // and our API is expecting linkmate to run
+    if (this.linkmate_ && this.linkmateOptions_.linkmateEnabled &&
+      this.linkmateOptions_.linkmateExpected) {
+      this.smartLinkRewriter_.getAnchorReplacementList();
+    }
+  }
+
+  /**
+   * API call to retrieve the Narrativ config for this extension.
+   * @return {?Promise}
+   * @private
+   */
+  getLinkmateOptions_() {
+    const fetchUrl = ENDPOINTS.NRTV_CONFIG_ENDPOINT.replace(
+        '.nrtv_slug.', this.linkmateOptions_.nrtvSlug
+    );
+
+    return this.xhr_.fetchJson(fetchUrl, {
+      method: 'GET',
+      ampCors: false,
+    })
+        .then(res => res.json())
+        .then(res => {
```

Comment:
Definitely will add a comment! If the return value is incorrect then the config values won't be set correctly and the API calls in `linkmate.js` will return 500. I think this is fine since anything that would cause that is a bad server side change by us and not the responsibility of `amp-smartlinks`.

Reply:
ah right. good catch. The same change should be made on the skim-links side. 

cc\ @slocka https://github.com/ampproject/amphtml/blob/master/extensions/amp-skimlinks/0.1/affiliate-link-resolver.js#L108

Thanks Yuxuan!